### DISCUSSION TITLE
Conditionally create a layer from a server response to a form submit

### DISCUSSION THREAD

**Discussion by joao-pedro-braz** [unpoly/unpoly#602] [Q&A] [answered]
Hello!

Trying to figure out a way I could conditionally create a new layer from a server's response to a form's submission, currently I have this form:

```html
<form method="POST"
              action="/login"
              class="mt-8 grid grid-cols-6 gap-6"
              x-data="{ termsAccepted: false }"
              up-submit>
...
</form>
```

And `up-submit` works wonders with regards to redirects and validation, but it will always replace the form's content with the servers response and I need to conditionally (if it's the first time the user's logging in) create a new layer (open a dialog) instead. Any ideas on how I could accomplish that?

**Comment by triskweline** [comment]
Here are two ideas. They involve setting a custom response header `First-Login: true` and reacting to that.

To overlay a dialog on top of the login success page, we could use this:

```js
up.on('up:fragment:loaded', ({ response }) => {
  if (response.header('First-Login')) up.layer.open({ url: '/welcome' })
})
```

If we must show show the dialog before proceeding to the login success page, we could react like this:

```js
up.on('up:fragment:loaded', ({ response, renderOptions }) => {
  if (response.header('First-Login')) renderOptions.layer = 'new'
})
```

**Reply by joao-pedro-braz** [reply]
Thank you for taking the time to answer my question :)

Using `renderOptions.layer` almost works, initially it complained I needed to also set `mode`, which I did, but then it failed to `find common target in current page and response`. That's probably related to the fact my server's just returning the fragment containing the dialog, instead of the whole page + dialog.

Event listener:
```js
up.on('up:fragment:loaded', ({ response, renderOptions }) => {
  if (response.header('X-Up-New-Layer')) {
    renderOptions.layer = 'new'
    renderOptions.mode = 'modal'
  }
})
```
Error: `Uncaught (in promise) up.CannotMatch: Could not find common target in current page and response (tried selectors [[up-main~=modal], [up-main~=overlay], [up-main=''], main] in layer "new")`

Any suggestions?

**Reply by triskweline** [reply, marked as answer]
The response needs to contain a main element or you need to set `renderOptions.target` to another selector.

**Reply by joao-pedro-braz** [reply]
Worked flawlessly.

Thank you so much!
